metacat-main/src/main/java/com/netflix/metacat/main/services/impl/TableServiceImpl.java

```diff
@@ -184,15 +189,48 @@ public TableDto deleteAndReturn(final QualifiedName name, final boolean isMView)
             return t;
         });
 
-        // Delete the metadata.  Type doesn't matter since we discard the result
-        log.info("Deleting user metadata for table {}", name);
-        userMetadataService.deleteMetadatas(metacatRequestContext.getUserName(), Lists.newArrayList(tableDto));
-        log.info("Deleting tags for table {}", name);
-        tagService.delete(name, false);
+        if (canDeleteMetadata(name)) {
+            // Delete the metadata.  Type doesn't matter since we discard the result
+            log.info("Deleting user metadata for table {}", name);
+            userMetadataService.deleteMetadatas(metacatRequestContext.getUserName(), Lists.newArrayList(tableDto));
+            log.info("Deleting tags for table {}", name);
+            tagService.delete(name, false);
+        } else {
+            if (config.canSoftDeleteDataMetadata() && tableDto.isDataExternal()) {
+                userMetadataService.softDeleteDataMetadatas(metacatRequestContext.getUserName(),
+                    Lists.newArrayList(tableDto.getDataUri()));
+            }
+        }
         eventBus.postAsync(new MetacatDeleteTablePostEvent(name, metacatRequestContext, this, tableDto, isMView));
         return tableDto;
     }
 
+    /**
+     * Returns true
+     * 1. If the system is configured to delete deifnition metadata.
+     * 2. If the system is configured not to but the tableName is configured to either explicitly or if the
+     * table's database/catalog is configure to.
+     * @param tableName table name
+     * @return whether or not to delete definition metadata
+     */
+    private boolean canDeleteMetadata(final QualifiedName tableName) {
+        return config.canDeleteTableDefinitionMetadata() || isEnabledForTableDefinitionMetadataDelete(tableName);
+    }
+
```

Comment:
Can we rename `canDeleteTableDefinitionMetadata()` as `forceTableDefinitionMetadataEnabled() `? looks like this flag forcing all the table defmetadata to be deleted.
and 
`isEnabledForTableDefinitionMetadataDelete(tableName)` as `tableDefinitionMetadataDeleteEnabled(tableName)` this flag is individual table flag

Reply:
I disagree with the rename. It could either be kept as canDeleteTableDefinitionMetadata or changed to isDeleteDefinitionMetadataEnabled. The second configuration is used as a whitelist for any qualified name.